src/metabase/models/secret.clj

```diff
@@ -25,7 +30,34 @@
           :can-write?        i/superuser?}))
 
 ;;; ---------------------------------------------- Hydration / Util Fns ----------------------------------------------
-;; none yet
+
+(defn value->string
+  "Returns the value of the given `secret` instance as a String."
+  {:added "0.42.0"}
+  ^String [{:keys [^bytes value] :as secret}]
+  (String. value StandardCharsets/UTF_8))
+
+(defn value->file!
+  "Returns the value of the given `secret` instance in the form of a file."
+  {:added "0.42.0"}
+  [{:keys [id ^bytes value] :as secret}]
+  (if (= :file-path (:source secret))
+    (let [secret-val          (value->string secret)
+          ^File existing-file (File. secret-val)]
+      (if (.exists existing-file)
+        existing-file
+        (throw (ex-info (tru "Secret {0} points to non-existent file: {1}" id secret-val)
+                 {:secret-id id
+                  :file-path secret-val}))))
+    (let [^File tmp-file (doto (File/createTempFile "metabase-secret_" nil)
```

Comment:
I'm not able to see how this is protected from data races. It seems to me that it could be possible to end up with several temp files created if there was some contention around this.

Reply:
Do you mean two threads both entering `value->file!` at the same time, or something else?  If the former, then it shouldn't be any problem (`createTempFile` creates a uniquely named file each time it's called).

Reply:
yes but shouldn't one create the file and then all others reuse that same file? I guess that fact makes me nervous on its own, but it makes me think of the connection pool restart process where maybe it sees a new file used and recreates the pool based on that file, and then it sees a different db with a different file and recreates with that file, etc.

Reply:
The same file will be used for all connections in a pool instance.  This ultimately gets called once in order to build up the JDBC spec, at which point the spec controls all N connections in the pool (thereby sharing the same path to a temp file).  If the connection details are changed, then this is invoked again, a new temp file generated, and all new connections in the updated pool will point to that.  I'm not sure what you mean by "it" in "it sees a new file used"?

Reply:
What is the mechanism that ensures this is only called once? And not three times at the same time? where is the lock?

Reply:
Right now, there isn't a lock.  The only "mechanism" really is the calling convention.  This particular PR doesn't show the full picture of how this gets used, unfortunately.  But you can see it in PR 6, [here](https://github.com/metabase/metabase/pull/18320/files#diff-71a39c38945d2c20a0214dc8fdbb14e8918bfdb66db304cf789bc7ecc6e21c90R93).  `handle-ssl-options` is ultimately called by `sql-jdbc.conn/connection-details->spec`, which I believe is only called by a single thread the first time a connection is needed.  That happens [here](https://github.com/metabase/metabase/blob/master/src/metabase/driver/sql_jdbc/connection.clj#L96), which **is** locked [here](https://github.com/metabase/metabase/blob/master/src/metabase/driver/sql_jdbc/connection.clj#L213-L219).

Now, of course, if the way in which we construct the connection pools per database instance ever changes, then this goes out the window.  But at least for now, I suppose I am leaning on the current call stack/convention for any sort of locking.